### ISSUE TITLE
The entity type AuditLog is not part of the model for the current context.

### ISSUE THREAD

**Issue by peters-patriotforge** [thepirat000/Audit.NET#450]
**Issue**

I intermittently get the following error:

`System.InvalidOperationException: The entity type AuditLog is not part of the model for the current context.`

I have difficulty reproducing the error and when I do, there doesn't seem to be any kind of pattern.

**Code**

DatabaseContext.cs

```csharp
public class DatabaseContext : AuditDbContext
{
        public DatabaseContext(string contextName) : base(contextName)
        {
            Configuration.LazyLoadingEnabled = false;
            Configuration.ProxyCreationEnabled = false;

            ConfigureAuditNet();
        }

        protected override void OnModelCreating(DbModelBuilder modelBuilder)
        {
            modelBuilder.Entity<AuditLog>().ToTable("AuditLogs");

            base.OnModelCreating(modelBuilder);       
        }

        private void ConfigureAuditNet()
        {
            Audit.Core.Configuration.Setup()
                .UseEntityFramework(_ => _
                    .AuditTypeMapper(t => typeof(AuditLog))
                    .AuditEntityAction<AuditLog>((ev, entry, entity) =>
                    {
                        entity.AuditData = entry.ToJson();
                        entity.EntityType = entry.EntityType.Name;
                        entity.AuditDate = DateTime.Now;
                        entity.AuditUserId = HttpContext.Current.User.GetUserIdInt();
                        entity.TablePk = entry.PrimaryKey.First().Value.ToString();
                    })
                    .IgnoreMatchedProperties(true));

            Audit.EntityFramework.Configuration.Setup()
                .ForContext<DatabaseContext>()
                .UseOptOut()
                    .Ignore<Heartbeat>();
        }

        public DatabaseContext() : base("name=OPv2Database") { }
 
       public DbSet<AuditLog> AuditLogs { get; set; }
       .
       .
       .
       // some more models
}
```

How I save to the database.

```csharp
public void AddShopOrderNotes(List<ShopOrderNote> shopOrderNotes)
{
            using (var context = new DatabaseContext(_connectionName))
            {
                context.ShopOrderNotes.AddRange(shopOrderNotes);

                context.SaveChanges();
            }
}
```

This will work anywhere from 80% - 95% of the time. The other times it will fail, then the user will save again immediately and it will work.

**Other Information**

`packages.config` in my project that defines the DatabaseContext
```xml
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="Audit.EntityFramework" version="15.1.1" targetFramework="net461" />
  <package id="Audit.NET" version="15.1.1" targetFramework="net461" />
  <package id="EntityFramework" version="6.2.0" targetFramework="net461" />
  <package id="Newtonsoft.Json" version="12.0.2" targetFramework="net461" />
</packages>
```
`packages.config` in my web api project
```xml
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="Antlr" version="3.5.0.2" targetFramework="net461" />
  <package id="Audit.EntityFramework" version="15.1.1" targetFramework="net461" />
  <package id="Audit.NET" version="15.1.1" targetFramework="net461" />
  <package id="AutoMapper" version="9.0.0" targetFramework="net461" />
  <package id="bootstrap" version="3.3.7" targetFramework="net461" />
  <package id="BouncyCastle" version="1.8.5" targetFramework="net461" />
  <package id="Destructurama.Attributed" version="2.0.0" targetFramework="net461" />
  <package id="Destructurama.ByIgnoring" version="1.0.4" targetFramework="net461" />
  <package id="EntityFramework" version="6.2.0" targetFramework="net461" />
  <package id="jQuery" version="3.3.1" targetFramework="net461" />
  <package id="MailKit" version="2.4.1" targetFramework="net461" />
  <package id="Microsoft.AspNet.Cors" version="5.2.7" targetFramework="net461" />
  <package id="Microsoft.AspNet.Mvc" version="5.2.4" targetFramework="net461" />
  <package id="Microsoft.AspNet.Razor" version="3.2.4" targetFramework="net461" />
  <package id="Microsoft.AspNet.TelemetryCorrelation" version="1.0.0" targetFramework="net461" />
  <package id="Microsoft.AspNet.Web.Optimization" version="1.1.3" targetFramework="net461" />
  <package id="Microsoft.AspNet.WebApi" version="5.2.4" targetFramework="net461" />
  <package id="Microsoft.AspNet.WebApi.Client" version="5.2.7" targetFramework="net461" />
  <package id="Microsoft.AspNet.WebApi.Core" version="5.2.7" targetFramework="net461" />
  <package id="Microsoft.AspNet.WebApi.Cors" version="5.2.7" targetFramework="net461" />
  <package id="Microsoft.AspNet.WebApi.HelpPage" version="5.2.4" targetFramework="net461" />
  <package id="Microsoft.AspNet.WebApi.WebHost" version="5.2.4" targetFramework="net461" />
  <package id="Microsoft.AspNet.WebPages" version="3.2.4" targetFramework="net461" />
  <package id="Microsoft.Bcl.AsyncInterfaces" version="1.1.0" targetFramework="net461" />
  <package id="Microsoft.CodeDom.Providers.DotNetCompilerPlatform" version="2.0.0" targetFramework="net461" />
  <package id="Microsoft.Extensions.Configuration" version="3.1.0" targetFramework="net461" />
  <package id="Microsoft.Extensions.Configuration.Abstractions" version="3.1.0" targetFramework="net461" />
  <package id="Microsoft.Extensions.Configuration.Binder" version="3.1.0" targetFramework="net461" />
  <package id="Microsoft.Extensions.DependencyInjection" version="3.1.0" targetFramework="net461" />
  <package id="Microsoft.Extensions.DependencyInjection.Abstractions" version="3.1.0" targetFramework="net461" />
  <package id="Microsoft.Extensions.Logging.Abstractions" version="3.1.0" targetFramework="net461" />
  <package id="Microsoft.Extensions.Options" version="3.1.0" targetFramework="net461" />
  <package id="Microsoft.Extensions.Primitives" version="3.1.0" targetFramework="net461" />
  <package id="Microsoft.IdentityModel.JsonWebTokens" version="5.4.0" targetFramework="net461" />
  <package id="Microsoft.IdentityModel.Logging" version="5.4.0" targetFramework="net461" />
  <package id="Microsoft.IdentityModel.Tokens" version="5.4.0" targetFramework="net461" />
  <package id="Microsoft.VisualStudio.SlowCheetah" version="3.2.26" targetFramework="net461" developmentDependency="true" />
  <package id="Microsoft.Web.Infrastructure" version="1.0.0.0" targetFramework="net461" />
  <package id="MimeKit" version="2.4.1" targetFramework="net461" />
  <package id="Modernizr" version="2.8.3" targetFramework="net461" />
  <package id="Newtonsoft.Json" version="12.0.2" targetFramework="net461" />
  <package id="Serilog" version="2.9.0" targetFramework="net461" />
  <package id="Serilog.Settings.AppSettings" version="2.2.2" targetFramework="net461" />
  <package id="Serilog.Sinks.File" version="4.1.0" targetFramework="net461" />
  <package id="SerilogWeb.Classic" version="5.0.48" targetFramework="net461" />
  <package id="SerilogWeb.Classic.WebApi" version="4.0.5" targetFramework="net461" />
  <package id="System.Buffers" version="4.4.0" targetFramework="net461" />
  <package id="System.ComponentModel.Annotations" version="4.7.0" targetFramework="net461" />
  <package id="System.Diagnostics.DiagnosticSource" version="4.4.1" targetFramework="net461" />
  <package id="System.IdentityModel.Tokens.Jwt" version="5.4.0" targetFramework="net461" />
  <package id="System.Memory" version="4.5.2" targetFramework="net461" />
  <package id="System.Numerics.Vectors" version="4.4.0" targetFramework="net461" />
  <package id="System.Runtime.CompilerServices.Unsafe" version="4.7.0" targetFramework="net461" />
  <package id="System.Threading.Tasks.Extensions" version="4.5.2" targetFramework="net461" />
  <package id="WebGrease" version="1.6.0" targetFramework="net461" />
</packages>
```

Please let me know if you need any other information. I'm stumped on what it could be and I'm willing to try any suggestions.

**Comment by thepirat000** [comment]
Consider calling your `ConfigureAuditNet` on the static constructor, or on your startup, so it's called only once

**Comment by peters-patriotforge** [comment]
Thanks for your response! I really appreciate it. After I made your suggested changes, I now get the same error every time I save to the database:

`System.InvalidOperationException: The entity type AuditLog is not part of the model for the current context.`

Here are the changes I made:

`DatabaseContext.cs`. I removed `ConfigureAuditNet()` from here.
```csharp
public class DatabaseContext : AuditDbContext
{
        public DatabaseContext(string contextName) : base(contextName)
        {
            Configuration.LazyLoadingEnabled = false;
            Configuration.ProxyCreationEnabled = false;
        }

        protected override void OnModelCreating(DbModelBuilder modelBuilder)
        {
            modelBuilder.Entity<AuditLog>().ToTable("AuditLogs");

            base.OnModelCreating(modelBuilder);       
        }

        public DatabaseContext() : base("name=OPv2Database") { }
 
       public DbSet<AuditLog> AuditLogs { get; set; }
       .
       .
       .
       // some more models
}
```

I created `AuditNetConfig.cs` and put my configuration code in a static `Configure()` method.
```csharp
public class AuditNetConfig
{
        public static void Configure()
        {
            Audit.Core.Configuration.Setup()
                .UseEntityFramework(_ => _
                    .AuditTypeMapper(t => typeof(AuditLog))
                    .AuditEntityAction<AuditLog>((ev, entry, entity) =>
                    {
                        entity.AuditData = entry.ToJson();
                        entity.EntityType = entry.EntityType.Name;
                        entity.AuditDate = DateTime.Now;
                        entity.AuditUserId = HttpContext.Current.User.GetUserIdInt();
                        entity.TablePk = entry.PrimaryKey.First().Value.ToString();
                    })
                    .IgnoreMatchedProperties(true));

            Audit.EntityFramework.Configuration.Setup()
                .ForContext<DatabaseContext>()
                .UseOptOut()
                    .Ignore<Heartbeat>();
        }
}
```

I then call it from my `Global.asax.cs` file at the bottom.
```csharp
public class WebApiApplication : System.Web.HttpApplication
{
    protected void Application_Start()
    {
        AreaRegistration.RegisterAllAreas();
        GlobalConfiguration.Configure(WebApiConfig.Register);
        FilterConfig.RegisterGlobalFilters(GlobalFilters.Filters);
        RouteConfig.RegisterRoutes(RouteTable.Routes);
        BundleConfig.RegisterBundles(BundleTable.Bundles);
        AutoMapperConfig.Configure();
        SerilogConfig.Configure();
        OPHttpClientConfig.Configure();
        AuditNetConfig.Configure();
    }
}
```

Am I missing something with configuring this globally? This would be my preferred way, but I couldn't get it to work.

**Comment by peters-patriotforge** [comment]
Actually it makes sense that it wouldn't work, because of the way I'm saving to the database:

```csharp
public void AddShopOrderNotes(List<ShopOrderNote> shopOrderNotes)
{
            using (var context = new DatabaseContext(_connectionName))
            {
                context.ShopOrderNotes.AddRange(shopOrderNotes);

                context.SaveChanges();
            }
}
```

I'm always creating a new instance and the new instance won't have the configuration anymore. At least that's how I understand it.

**Comment by thepirat000** [comment]
Are your Entities and DbContext defined on different assemblies?

Check this: https://stackoverflow.com/questions/20688922/the-entity-type-type-is-not-part-of-the-model-for-the-current-context

**Comment by peters-patriotforge** [comment]
Yes, my `AuditLog.cs` and `DatabaseContext.cs` are both in a "Models" project. Which I pull into a "WebApi" project, where my `Global.asax.cs` configures Audit.NET.

I read everything in the SO link you posted, but I'm doing what the accepted answer says and the other answers don't apply.

```csharp
protected override void OnModelCreating(DbModelBuilder modelBuilder)
{
    modelBuilder.Entity<AuditLog>().ToTable("AuditLogs");

    base.OnModelCreating(modelBuilder);       
}
```

I'm not sure what else I'm missing. Unless I'm misunderstanding what you're telling me.

**Comment by thepirat000** [comment]
As a test, can you check what happens if you try to insert an AuditLog entity to the DbSet of your DbContext programmatically and then call SaveChanges()?

**Comment by peters-patriotforge** [comment]
I got the same error, `System.InvalidOperationException: The entity type AuditLog is not part of the model for the current context.`

When I inserted the AuditLog entity programmatically :
```csharp
using (var context = new DatabaseContext(_connectionName))
{
    context.AuditLogs.Add(new AuditLog { AuditData = "Test 1" });

    context.SaveChanges();
}
```

**Comment by thepirat000** [comment]
you also need to ignore the AuditLog entity type when using OptOut. Try adding .Ignore<AuditLog>

**Comment by peters-patriotforge** [comment]
Now it saves and I don't get any errors after doing this:
```csharp
Audit.EntityFramework.Configuration.Setup()
                .ForContext<DatabaseContext>()
                .UseOptOut()
                    .Ignore<AuditLog>()
                    .Ignore<Heartbeat>();
```

**Comment by peters-patriotforge** [comment]
I meant to say that I don't get any errors when I do the following with the changes you suggested:

```c#
using (var context = new DatabaseContext(_connectionName))
{
    context.AuditLogs.Add(new AuditLog { AuditData = "Test 1" });

    context.SaveChanges();
}
```
Unfortunately, when I try to add any other data, like this:
```c#
public void AddShopOrderNotes(List<ShopOrderNote> shopOrderNotes)
{
            using (var context = new DatabaseContext(_connectionName))
            {
                context.ShopOrderNotes.AddRange(shopOrderNotes);

                context.SaveChanges();
            }
}
```

I still get the following error:

`System.InvalidOperationException: The entity type AuditLog is not part of the model for the current context.`

This makes me think that it still doesn't recognize what I am setting up in my `Global.asax.cs` file.

**Comment by thepirat000** [comment]
Can you try by decorating the `AuditLog` class with `[AuditIgnore]` attribute instead of using `Ignore<AuditLog>()` ?

Can you share a minimal project that reproduces the issue?

**Comment by peters-patriotforge** [comment]
Following your advice, I have set up this [project](https://github.com/peters-patriotforge/audit-net-global-config) that demonstrates what I am trying to do. While setting this project up, I realized where I had went wrong and why trying to fix it didn't work.

**Current Situation**

At the moment, I use Audit.NET in two different databases (I left this part out, because I was so focused on the one database that I was getting issues with). I am able to have this working most of the time, because I set up the configuration in the `DatabaseContext.cs` file and the `SecurityDatabaseContext.cs` file. They are set up in the same way:

`DatabaseContext.cs`
```c#
public class DatabaseContext : AuditDbContext
{
        public DatabaseContext(string contextName) : base(contextName)
        {
            Configuration.LazyLoadingEnabled = false;
            Configuration.ProxyCreationEnabled = false;

            ConfigureAuditNet();
        }

        protected override void OnModelCreating(DbModelBuilder modelBuilder)
        {
            modelBuilder.Entity<AuditLog>().ToTable("AuditLogs");

            base.OnModelCreating(modelBuilder);       
        }

        private void ConfigureAuditNet()
        {
            Audit.Core.Configuration.Setup()
                .UseEntityFramework(_ => _
                    .AuditTypeMapper(t => typeof(AuditLog))
                    .AuditEntityAction<AuditLog>((ev, entry, entity) =>
                    {
                        entity.AuditData = entry.ToJson();
                        entity.EntityType = entry.EntityType.Name;
                        entity.AuditDate = DateTime.Now;
                        entity.AuditUserId = HttpContext.Current.User.GetUserIdInt();
                        entity.TablePk = entry.PrimaryKey.First().Value.ToString();
                    })
                    .IgnoreMatchedProperties(true));

            Audit.EntityFramework.Configuration.Setup()
                .ForContext<DatabaseContext>()
                .UseOptOut()
                    .Ignore<Heartbeat>();
        }

        public DatabaseContext() : base("name=OPv2Database") { }
 
       public DbSet<AuditLog> AuditLogs { get; set; }
       .
       .
       .
       // some more models
}
```

`SecurityDatabaseContext.cs`
```c#
// Same as set up as DatabaseContext
.
.
.

public SecurityDatabaseContext() : base("name=OPv2SecurityDatabase") { }
 
       public DbSet<AuditLog> AuditLogs { get; set; }
       .
       .
       .
       // some more models
```

This worked for the most part. The error, `The entity type AuditLog is not part of the model for the current context.`, would come up once and a blue moon. But, as usage became higher, I started seeing it more often. It's still not that much, but enough to be annoying.

Essentially it looks like the configurations would constantly get changed based on which context was used. This makes sense, why the error was intermittent when it was set up this way.

Once I changed it globally, then only one or the other would work perfectly. So my problem turned out to be setting up multiple ef contexts with Audit.NET.

**Solution**

I ended up finding your answer to another issue #435 which you advised to create a custom DataProvider, called `MultiEntityFrameworkDataProvider`, which worked perfectly.

```c#
public class AuditNetConfig
    {
        public static void Configure()
        {
            Audit.Core.Configuration.Setup()
                .UseCustomProvider(new MultiEntityFrameworkDataProvider());

            var dp = Audit.Core.Configuration.DataProvider as MultiEntityFrameworkDataProvider;

            dp.AttachDataProvider<DatabaseContext>(new EntityFrameworkDataProvider(ef => ef
                .UseDbContext<DatabaseContext>()
                    .AuditTypeMapper(t => typeof(Models.AuditLog))
                    .AuditEntityAction<Models.AuditLog>((ev, entry, entity) =>
                    {
                        entity.AuditData = entry.ToJson();
                        entity.EntityType = entry.EntityType.Name;
                        entity.AuditDate = DateTime.Now;
                        entity.TablePk = entry.PrimaryKey.First().Value.ToString();
                    })
                    .IgnoreMatchedProperties(true)));
            
            dp.AttachDataProvider<SecurityDatabaseContext>(new EntityFrameworkDataProvider(ef => ef
                .UseDbContext<SecurityDatabaseContext>()
                    .AuditTypeMapper(t => typeof(Security.Models.AuditLog))
                    .AuditEntityAction<Security.Models.AuditLog>((ev, entry, entity) =>
                    {
                        entity.AuditData = entry.ToJson();
                        entity.EntityType = entry.EntityType.Name;
                        entity.AuditDate = DateTime.Now;
                        entity.TablePk = entry.PrimaryKey.First().Value.ToString();
                    })
                    .IgnoreMatchedProperties(true)));
        }
    }
```

This is the code from [AuditNetConfig.cs](https://github.com/peters-patriotforge/audit-net-global-config/blob/main/AuditNetGlobalConfig.WebApi/App_Start/AuditNetConfig.cs), which I shameless lifted from your previous work :)

Thank you for all your help and patience. This is a top notch open source project and you're an excellent project owner!

**Comment by thepirat000** [comment]
That's the way to go. The data provider should be set just once, since it's a global static property (`Audit.Core.Configiration.DataProvider`)

And thank you very much for your kind words :)
